r-package/grf/R/input_utilities.R

```diff
@@ -220,18 +220,45 @@ validate_sample_weights <- function(sample.weights, X) {
 
 #' @importFrom Matrix Matrix cBind
 #' @importFrom methods new
-create_data_matrices <- function(X, ..., sample.weights = NULL) {
+create_data_matrices <- function(X, outcome = NULL, treatment = NULL,
+                                 instrument = NULL, sample.weights = FALSE) {
   default.data <- matrix(nrow = 0, ncol = 0)
   sparse.data <- new("dgCMatrix", Dim = c(0L, 0L))
+  out <- list()
+  i <- 1
+  if (!is.null(outcome)) {
+    out[["outcome_index"]] <- ncol(X) + i
+  }
+  if (!is.null(treatment)) {
+    i <- i + 1
+    out[["treatment_index"]] <- ncol(X) + i
+  }
+  if (!is.null(instrument)) {
+    i <- i + 1
+    out[["instrument_index"]] <- ncol(X) + i
+  }
+  if (!isFALSE(sample.weights)) {
```

Comment:
Why is `sample.weights=FALSE` by default, as opposed to `sample.weights=NULL`?
 
What's the desired behavior for `sample.weights=TRUE`? (Right now, it appends a column of `1`s)

Reply:
it is convoluted (a better way?) because of how sample.weights is handled

FALSE is to distinguish 1 vs 2:

1. passing on to regression_train and causal_train:

even when sample.weight = NULL: sample_weight_index is required in the bindings
(together with use_sample_weights)

1.1: If you are calling these: call create_data_matrices with NULL if no sample.weights, else with the sample.weights specified 

2. passing on to for ex. regression_predict: does not take sample.weight, and does
not require sample_weight_index as above: specify this with sample.weights = FALSE

Reply:
Maybe a simplification is to remove [bool use_sample_weights](https://github.com/grf-labs/grf/blob/master/r-package/grf/bindings/RegressionForestBindings.cpp#L36)

and instead inside do bool use_sample_weights = sample_weight_index > 0

and in create_data_matrices: set sample_weight_index = 0 if NULL